core/src/swarm/registry.rs

```diff
@@ -62,17 +66,41 @@ impl Addresses {
     /// Adding an existing address is interpreted as additional
     /// confirmation and thus increases its score.
     pub fn add(&mut self, a: Multiaddr) {
+        debug_assert_ne!(self.reports.capacity(), 0);
+
+        let oldest = if self.reports.len() == self.reports.capacity() {
+            self.reports.pop_front()
+        } else {
+            None
+        };
+
+        if let Some(oldest) = oldest {
+            if let Some(in_registry) = self.registry.iter_mut().find(|r| r.addr == oldest) {
+                in_registry.score = in_registry.score.saturating_sub(1);
+                isort(&mut self.registry);
+            }
+        }
+
+        // Remove addresses that have a score of 0.
```

Comment:
I suggest not to remove records with score 0 here. The important goal is that outdated addresses move to the end quickly where they can still be left lingering before being pushed out by newer addresses. There might still be a chance that they are reachable.

Reply:
Addresses with a score of 0 haven't been reported in the last 200 reports.
Removing addresses only when we reach the limit of 30 addresses assumes that your address actually rotates often enough, which might not necessarily be true.

If your address rotates only once every three months, you don't want to wait for 900 months in order for the oldest address to disappear.

Reply:
The current address would take precedence long before that so the fact that there is this outdated address after the current one(s) is of no concern and with a limited report history a seemingly unreported address may after all still be useful.

Reply:
We're still reporting to everyone an extra address which we know is useless.

Reply:
The code has no notion of time beyond the limited historical record so it can only be *assumed* that an address with score 0 is "useless".

Reply:
I don't really get the point. The fact that we're limiting ourselves to 30 addresses is also extremely arbitrary, and we can't be sure that the 31st address is not actually valid.

There's no perfect solution to that, and I'm just trying to do heuristics to solve the problem of rotating IPs and bad/malicious reports.

Reply:
My point is that because we already have this arbitrary limit we do not need another one. It is there as a safeguard to ensure the number of addresses does not grow unbounded and certainly does not imply an address is useless as it will remove addresses even if their score is > 0. The report history on the other hand is an orthogonal mechanism to determine the order of addresses based on historic information but it can not judge the usefulness of addresses beyond its historical record.

Reply:
This PR kind of suggests to entirely change the mechanism so that the list of addresses **is** the historical reports of addresses. Having this limit of 30 addresses would then become a "secondary check" and could maybe be put inside the logic of `AddressesIter` instead.

Reply:
> [...] the list of addresses **is** the historical reports of addresses [...]

In that case I would remove the registry limit altogether. The max. history length already determines a max. registry limit if we remove records with score 0. Client code using the iterator can then just `take` an arbitrary-length prefix of that (e.g. 30).